### ISSUE TITLE
Is it possible to do online learning with a stateful LSTM?

### ISSUE THREAD

**Issue by zacheberhart** [keras-team/keras#7596]
I have an ensemble of LSTMs and the training time is currently very high, about several hours. Unfortunately, I need to re-train the models fairly often which obviously takes a lot of time. That said, I wanted to try to cut out some of the training time by saving the weights and such but I'm worried that it isn't possible with a stateful LSTM (mainly because I'm not shuffling the samples).

[I was reading this](https://github.com/fchollet/keras/issues/1868) and it seems that there a few issues so I wanted to get some clarification.

This is roughly the process I am thinking:

1. train / fit a stateful LSTM (1 timestep) on a dataset (time series) with a batch size of 1 and shuffling set to False.
2. export model
3. after a day or so I have several more samples
4. import the model
5. train model on additional data
6. repeat

Things I'm worried about are: the restarting of the learning rate each time `.fit()` is called and the new data overwriting what the model has learned (as per the issue referenced above).

Is this feasible?

**Comment by zacheberhart** [comment]
I still haven't found a solution to this. Anyone have a similar problem? Or a solution?

**Comment by MrMathias** [comment]
The worries you mention are inherent to online learning. If your domain changes over time, the learning rate will be a way to tweak how myopic your model is.

Regarding the optimization state resetting, Keras maintain this using the mode.save() call:

> You can use model.save(filepath) to save a Keras model into a single HDF5 file which will contain:
> 
> * the architecture of the model, allowing to re-create the model
> * the weights of the model
> * the training configuration (loss, optimizer)
> * **the state of the optimizer, allowing to resume training exactly where you left off.**

I would double check if the state of a stateful LSTM is stored as well, but following the docs, it seems to be the case.

**Comment by Metalkiler** [comment]
Hi, according to online learning I believe the issue is something like this:
Today we get a dataset with 3 diferent values (e.g, A,B,C) if at any given moment we got a new letter (named D) we could just update the model...

For instance I believe that online learning keeps the weights of the hidden layers and "dumps" the input layers and the output layer (since it will be different) so the code I developed should help with online learning :) (Ps: X is a sparse matrix )



#para atualizar o modelo com novos dados é preciso:
#deitar fora o primeiro layer e o ultimo layer
#os restantes layers ficam com os pesos precisos..
model2 = Sequential()
model2.add(Dense(1000, input_dim=len(train_X.columns), activation='relu', name='new_Inputs'))
 
for layer in model.layers[1:-1]:
    model2.add(layer)

model2.add(Dense(1, activation='sigmoid'))
 
model2.compile(loss='binary_crossentropy',
              optimizer='AdaDelta',
              metrics=['accuracy'])

history=model2.fit( X , train_Y.target,
              epochs=10,
              batch_size=2**10,callbacks = callbacks_list,validation_split=0.2)


Do you agree with this ideia as an way to update the model with volatile data (such is the case with Big Data) ?
